modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java

```diff
@@ -266,22 +266,36 @@ else if (type instanceof UuidType)
                 return createTypeWithNullability(createSqlIntervalType(INTERVAL_QUALIFIER_DAY_TIME), true);
             else if (clazz == Period.class)
                 return createTypeWithNullability(createSqlIntervalType(INTERVAL_QUALIFIER_YEAR_MONTH), true);
-            else if (clazz == UUID.class)
-                return createTypeWithNullability(createUuidType(), true);
+            else {
+                RelDataType relType = createCustomType(clazz.getName());
+
+                if (relType != null)
+                    return relType;
+            }
         }
 
         return super.toSql(type);
     }
 
-    /** @return UUID SQL type. */
-    public RelDataType createUuidType() {
-        return canonize(new UuidType(true));
+    /** @return Custom type by name or storeable class name. {@code Null} if type not found. */
+    public RelDataType createCustomType(String typeName) {
+        return createCustomType(typeName, true);
+    }
+
+    /** @return Nullable custom type by name or storeable class name. {@code Null} if type not found. */
+    public RelDataType createCustomType(String typeName, boolean nullable) {
+        if ("UUID".equals(typeName) || UUID.class.getName().equals(typeName))
+            return new UuidType(nullable);
+        else if ("OTHER".equals(typeName) || Object.class.getName().equals(typeName))
+            return new OtherType(nullable);
+
+        return null;
     }
 
     /** {@inheritDoc} */
     @Override public RelDataType createTypeWithNullability(RelDataType type, boolean nullable) {
-        if (type instanceof UuidType && type.isNullable() != nullable)
-            type = new UuidType(nullable);
+        if (type instanceof IgniteSqlCalciteType && type.isNullable() != nullable)
```

Comment:
`createType` method also should be modified (`UUID` and `Object` classes should be added, and probably `byte[]` for BINARY type). To reproduce the problem without modifying `createType()` method you can add new column to the `JdbcQueryTest#testQueryColumnTypes` test, for example:
```
            "binary_col BINARY, " +
            "uuid_col UUID, " +
            "other_col OTHER, " +
```
And check:
```
            assertEquals(Types.BINARY, md.getColumnType(18));
            assertEquals(byte[].class.getName(), md.getColumnClassName(18));
            assertEquals(Types.OTHER, md.getColumnType(19));
            assertEquals(UUID.class.getName(), md.getColumnClassName(19));
            assertEquals(Types.OTHER, md.getColumnType(20));
            assertEquals(Object.class.getName(), md.getColumnClassName(20));
```
Currently, all these columns return `Types.OTHER`  as column type and `Object[]` as column class name